### ISSUE TITLE
Travis improvements

### ISSUE THREAD

**Issue by Psirus** [nutofem/nuto#64]
Speed
-----

The Travis builds take quite a while, especially the coverage/debug build is now up to almost 28 minutes. Especially for pull requests, where each commit is build twice (once on the branch, once "as if" merged into master), the wait can become quite long. Also, I think that we should strive to be as economical as possible in our use of infrastructure that has been provided to us for free.

For example, just as locally, one can [use ccache with Travis](https://docs.travis-ci.com/user/caching).

Another option is to [use Docker images](https://docs.travis-ci.com/user/docker/). While I'm no expert, the idea is to bundle everything you need into one container. So instead of running `apt-get install very many packages` every build, you create a sort of image that has all the dependencies and use that instead. I'm not sure if this will be quicker, but maybe someone wants to try.

Seperate Unit/Integration Coverage
----------------------------------

With the use of [flags](https://docs.codecov.io/docs/flags), the code coverage can be seperated into results from unit tests and results from integration tests. Imho this would be really interesting to see. Also, one could seperate the coverage of `src/` from the coverage of the tests themselves. Take a look at the [manpage of lcov](http://ltp.sourceforge.net/coverage/lcov/lcov.1.php) on how to use the `--remove` and `--extract` options to achieve this. I suggest trying this locally first ([lcov quickstart guide](http://ltp.sourceforge.net/coverage/lcov.php), instead of many iterations on Travis.

**Comment by phuschke** [comment]
**Status quo**
I managed to pull a docker image and use that image to run make + ctest. I tested it on this simple cpp project if you want to have a look:

[phuschke/GithubCppProjectTemplate](https://github.com/phuschke/GithubCppProjectTemplate)

Will build a new docker image for NuTo and update .travis.yml soon.

**Comment by TTitscher** [comment]
I like! Any approximations for the speedup?

**Comment by TTitscher** [comment]
I tried to go on a different road, following https://atchai.com/blog/faster-ci/

Travis is somehow able to cache our apt-get packages. This sounds like, what we want. But: We have to switch to their [APT Addon](https://docs.travis-ci.com/user/installing-dependencies/#Installing-Packages-on-Container-Based-Infrastructure). The _only_ problem I currently have: `swig3.0`. This is included in ubuntu/trusty-backports, which is no problem for our current travis config. With this APT Addon thing, trusty-backports is not included. In addition, we cannot use `sudo` if we want to do that... Any ideas?

Any progress on Docker, @phuschke ? Can I help you somehow?

EDIT: The procedure in the link above does not actually cache the apt-get stuff. I misread that. Gogo Docker. AFAIK, the current problem is the upload of our coverage stuff from inside the docker image to the website.

**Comment by TTitscher** [comment]
@phuschke and I worked on the docker stuff for a while. You can see the results [here](https://github.com/nutofem/nuto/blob/docker/.travis.yml). 
We created a [docker container](https://hub.docker.com/r/nuto/nuto_docker/tags/) that contains Ubuntu16.04 and all the `apt-get` stuff we need. On each travis run, this image is pulled instead of running `apt-get` each time. __This saves _~2 minutes_ per travis run.__ Not too bad.

On top of that, we no longer require sudo rights in travis, which has some [benefits](https://docs.travis-ci.com/user/reference/precise/#Virtualization-environments) (or are they drawbacks?, who knows...)

However, it is not fully working. Remaining issues:

- [x]  `docker exec` runs with sudo rights. MPI says:
~~~
mpirun has detected an attempt to run as root.
Running at root is *strongly* discouraged as any mistake (e.g., in
defining TMPDIR) or bug can result in catastrophic damage to the OS
file system, leaving your system in an unusable state.
You can override this protection by adding the --allow-run-as-root
option to your cmd line. However, we reiterate our strong advice
against doing so - please do so at your own risk.
~~~

- [ ] some issues regarding input files like
~~~
Error opening input file MeshCompanionGmshFine.msh for read access.
~~~

- [x] upload of the coverage. We did follow the [instructions](https://docs.codecov.io/docs/testing-with-docker) but still did not manage to do it. c003dcf7643e245583f3f59d72c563d9b56b5d84 is our commit, [this](https://travis-ci.org/nutofem/nuto/builds/260949397?utm_source=github_status&utm_medium=notification) is the travis output. At the very bottom you see 
~~~
bash: /dev/fd/63: No such file or directory
~~~

We want to keep our sanity and stop this project for now. Maybe someone else has the strength, willpower and (first and foremost) _luck_ to continue this glorious adventure. :ghost:

**Comment by Psirus** [comment]
Recently, the number of workarounds and manual install we do on Travis is increasing, mainly because of the outdated Ubuntu version. Travis has [no immediate plans to upgrade](https://github.com/travis-ci/travis-ci/issues/5821). Therefore we would like to pick this up again in the future. Until then, I want to collect the areas that will get cleaned up.

- [x] [`enum_hack.py`](https://github.com/nutofem/nuto/blob/master/python/nuto/enum_hack.py)
- [x] [`install_xyz.sh`](https://github.com/nutofem/nuto/tree/master/scripts)
- [x] [`ppa:ubuntu-toolchain-r`](https://github.com/nutofem/nuto/blob/master/.travis.yml)
- [ ] clang + openmp (#105)

**Comment by Psirus** [comment]
Since #158 has been merged, the coverage is a bit "messed up". For one, the `src/base/Version.cpp` is not filed into the proper directory. Also, the coverage of fakeit is wrongly included.
[Before](https://codecov.io/gh/nutofem/nuto/tree/a723474d773b763d68eb3fd0d7460d998a4c9768)
[After](https://codecov.io/gh/nutofem/nuto/tree/5482c8c97576bfae8080eddddb38bf117226f397)

**Edit:** Fixed in #167.
